Optimize compact reconstruction somewhat

TheBlueMatt opened the pull request:
This is some stuff I did for FIBRE a while back (that @rustyrussell cleaned up a lot, thanks!) and figured could get upstreamed. It saves some milliseconds when reconstructing, but for upstream its not all that critical.

TheBlueMatt committed:
Use open_hash_set in blockencodings

TheBlueMatt committed:
Add an implementation of an open hash set

rustyrussell committed:
open_hash_set: use a smaller (and power of two) hash table.

This makes performance a little worse, due to more hash collisions,
depending on how sparse we make it.

Since we're going to switch to linear chaining, we can afford slightly
more collisions, sowe drop to 4x.

50% full (2x oversize):
	Total insert hashes = 1977, total find hashes = 87224

25% full (4x oversize):
	Total insert hashes = 1749, total find hashes = 63927

12.5% full (8x oversize, equivalent to original):
	Total insert hashes = 1639, total find hashes = 56644

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
open_hash_set: use a linear hash scan.

No more jumping around the table, and just use the lower bits as hash.

	Total insert hashes = 1762, total find hashes = 65245

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
open_hash_set: make scan_max a constant.

For a given density, the chances of more than N collisions is
indendent of hash table size.  We chose 1/4 density, so we can assert
that the chances of 20 collisions is 1 in 2^40.

FECHeaderRTTTest1550,288,0.003517217934132,0.003554463386536,0.003529562718338,9538257,9639290,9571711

TheBlueMatt committed:
Return early in PartiallyDownloadedBlock if we got full-prefill

TheBlueMatt committed:
Dont alloc twice in block header deserialize

gmaxwell commented on src/blockencodings.cpp:
```diff
@@ -131,13 +156,16 @@ ReadStatus PartiallyDownloadedBlock::InitData(const CBlockHeaderAndShortTxIDs& c
-                txn_available[idit->second] = extra_txn[i].second;
-                have_txn[idit->second]  = true;
+        uint64_t next_shortid = 0;
+        if (i + 1 < extra_txn.size())
```
unbraced

TheBlueMatt replied:
Oops, forgot to style the stuff, fixed now.

arvidn commented on src/blockencodings.cpp:
```diff
@@ -43,6 +44,32 @@ uint64_t CBlockHeaderAndShortTxIDs::GetShortID(const uint256& txhash) const {
+namespace {
+    struct ShortIdIndexPair {
+        uint64_t shortid : 48;
+        uint16_t index : 16;
```
in order to make these packed together, you should change ``uint16_t`` to ``uint64_t``. Otherwise these are two separate bitfields with different allocation units.

TheBlueMatt replied:
Are you sure? There's a static_assert on the sizeof on the next line that doesnt make compile fail?

arvidn replied:
https://travis-ci.org/bitcoin/bitcoin/jobs/256256286#L2508

arvidn replied:
I think compilers are free to pack them still if they want to, but my recollection is that they aren't required to

TheBlueMatt replied:
Oh, funny, didnt realize it failed on 32 bit windows, but passed on 64-bit linux.

arvidn commented on src/open_hash_set.h:
```diff
@@ -0,0 +1,106 @@
+private:
+    hasher hash_instance;
+    key_equal equal_instance;
+    IsKeyNull null_instance;
```
all these instances are going to occupy one byte of storage each, despite not having any state (this is because of the guarantee C++ give you that every object has a unique address). It may not matter, but one trick to make them not use any space is to derive from them, and rely on the empty base class optimization

TheBlueMatt replied:
Hmm, if I'm understanding your suggestion correctly I find the current code much more readable, and the few extra bytes isnt worth a readabilitiy loss.

arvidn replied:
makes sense

rustyrussell committed:
open_hash_set: don't bother counting scan_max in find().

The caller aborts when an insert fails (as expected), so don't need
to check here, nor the extra equal_instance() check.

	#Benchmark,count,min,max,average,min_cycles,max_cycles,average_cycles
	FECHeaderRTTTest1550,320,0.003286004066467,0.003375306725502,0.003304844349623,8911484,9153595,8962469

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

rustyrussell committed:
open_hash_set: make find_fast() that uses a const table.

I cheated here, and just returned a damn pointer.  Similar gains might be
made from creating a const_iterator, but end() is not NULL, so it's a more
expensive comparison in the caller.

	#Benchmark,count,min,max,average,min_cycles,max_cycles,average_cycles
	FECHeaderRTTTest1550,320,0.003265343606472,0.003385126590729,0.003300794214010,8855479,9180430,8951641

Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>

MeshCollider commented on src/blockencodings.cpp:
```diff
@@ -43,6 +44,32 @@ uint64_t CBlockHeaderAndShortTxIDs::GetShortID(const uint256& txhash) const {
     return SipHashUint256(shorttxidk0, shorttxidk1, txhash) & 0xffffffffffffL;
 }
 
+namespace {
```
Nit, namespace brace on same line

MeshCollider commented on src/blockencodings.cpp:
```diff
@@ -43,6 +44,32 @@ uint64_t CBlockHeaderAndShortTxIDs::GetShortID(const uint256& txhash) const {
+            return elem.shortid == 0 && elem.index == 0;
+        }
+    };
+};
```
Semicolon, and should have end of namespace comment according to style guide

MeshCollider commented on src/open_hash_set.h:
```diff
@@ -0,0 +1,107 @@
+/** Implements an open hash set.
+ */
+template<class Key, class IsKeyNull, class Hash = std::hash<Key>, class KeyEqual = std::equal_to<Key>>
+class open_hash_set {
```
Nit, brace

MeshCollider commented on src/open_hash_set.h:
```diff
@@ -0,0 +1,107 @@
+    typedef Hash hasher;
+    typedef KeyEqual key_equal;
+
+    class iterator {
```
Brace

meshcollider reviewed (commented):
A few more style nits, also all the function braces too

promag commented on src/open_hash_set.h:
```diff
@@ -0,0 +1,107 @@
+
+#include <utility>
+
+#include <stdio.h>
```
Remove.

promag commented on src/open_hash_set.h:
```diff
@@ -0,0 +1,107 @@
+#ifndef OPEN_HASH_MAP_H
+#define OPEN_HASH_MAP_H
+
+#include <utility>
```
Use these instead:
```cpp
#include <functional>
#include <vector>
```

TheBlueMatt replied:
Need utility for pair.

promag reviewed (commented):
Is this going forward? I mean, according to @TheBlueMatt, the benefits aren't that much and requires to maintain `open_hash_set`.

TheBlueMatt commented:
@promag definitely not for 15, so lets wait to see what people think when the focus is back on other stuff.
